[HUDI-9405][DNM] Adding end to end streaming writes to metadata table support for SPARK engine

nsivabalan opened the pull request:
### Change Logs

_Describe context and summary for this change. Highlight if any code was copied._

### Impact

_Describe any public API or user-facing feature change or any performance impact._

### Risk level (write none, low medium or high below)

_If medium or high, explain what verification was done to mitigate the risks._

### Documentation Update

_Describe any necessary documentation update if there is any new feature, config, or user-facing change. If not, put "none"._

- _The config description must be updated if new configs are added or the default value of the configs are changed_
- _Any new feature or user-facing change requires updating the Hudi website. Please create a Jira ticket, attach the
  ticket number here and follow the [instruction](https://hudi.apache.org/contribute/developer-setup#website) to make
  changes to the website._

### Contributor's checklist

- [ ] Read through [contributor's guide](https://hudi.apache.org/contribute/how-to-contribute)
- [ ] Change Logs and Impact were stated clearly
- [ ] Adequate tests were added if applicable
- [ ] CI passed

nsivabalan commented on hudi-client/hudi-spark-client/src/test/java/org/apache/hudi/io/KeyGeneratorForDataGeneratorRecords.java:
```diff
@@ -0,0 +1,41 @@
+
+import org.apache.avro.generic.GenericRecord;
+
+public class KeyGeneratorForDataGeneratorRecords extends BuiltinKeyGenerator {
```
This was part of https://github.com/apache/hudi/pull/13286. This patch is stacked on top of that.

nsivabalan commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/BaseHoodieWriteClient.java:
```diff
@@ -239,13 +239,23 @@ public boolean commitStats(String instantTime, List<HoodieWriteStat> stats,
+                             String commitActionType, Map<String, List<String>> partitionToReplaceFileIds,
+                             Option<BiConsumer<HoodieTableMetaClient, HoodieCommitMetadata>> extraPreCommitFunc,
+                             boolean skipStreamingWritesToMetadataTable,
+                             Option<HoodieTable> hoodieTableOpt) {
```
Note to Reviewer: 
skipStreamingWritesToMetadataTable : this is mainly for row writer flows. With row writer, we can't really do the rdd chaining all the way since custom data source writes will return an array of WriteStatus and there is no way we can control the rdd or the triggering of writes. 

So, we had to introduce this argument named "skipStreamingWritesToMetadataTable". So that following call stack will avoid streaming writes flow and take the legacy write flow. 

We do have logic below to enable or disable based on write operations. Just that we have bulk insert support in both rdd based write client and also based on row writer. So, I am taking this route so that we keep streaming writes disabled for row writer paths. 

Happy to jam on any better ideas on this.

nsivabalan commented on hudi-spark-datasource/hudi-spark-common/src/main/java/org/apache/hudi/internal/DataSourceInternalWriterHelper.java:
```diff
@@ -83,8 +84,9 @@ public void onDataWriterCommit(String message) {
-          CommitUtils.getCommitActionType(operationType, metaClient.getTableType()));
+      writeClient.commitStats(instantTime, writeStatList, Collections.emptyList(), Option.of(extraMetadata),
+          CommitUtils.getCommitActionType(operationType, metaClient.getTableType()), Collections.emptyMap(), Option.empty(),
+          true, Option.empty());
```
Note to Reviewer:
here is where we set `skipStreamingWritesToMetadataTable` to true.

nsivabalan commented on hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/functional/RecordLevelIndexTestBase.scala:
```diff
@@ -131,6 +131,9 @@ class RecordLevelIndexTestBase extends HoodieStatsIndexTestBase {
       val recordLocation = recordIndexMap.get(recordKey)
+      if (recordLocation == null) {
+        println("afsdf")
+      }
```
fixed in latest commit pushed.

nsivabalan commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java:
```diff
@@ -1096,6 +1119,126 @@ public void buildMetadataPartitions(HoodieEngineContext engineContext, List<Hood
+  public HoodieData<WriteStatus> streamWriteToMetadataPartitions(HoodieData<WriteStatus> writeStatus, String instantTime) {
+    List<MetadataPartitionType> mdtPartitionsToTag = new ArrayList<>(enabledPartitionTypes);
+    mdtPartitionsToTag.remove(FILES);
+    mdtPartitionsToTag.retainAll(STREAMING_WRITES_SUPPORTED_PARTITIONS);
```
Note to reviewer:
In this patch we are only adding RLI partition to streaming flow. All other partitions will take up the 2nd write to metadata table i.e. non streaming writes. 
Overall, both writes to mdt put together will ensure all enabled partitions are written to MDT. 

In future, we will move 1 partition from non streaming write to streaming write flow. This is designed this way, so that we don't need to drag the patch until all partitions are added to streaming way of writing to mdt. 

And also, from inconsistency standpoint, its only RLI and Secondary index thats of very critical that definitely needs to be moved to streaming manner. All others are solely reliant on HoodieCommitMetadata which does not really matter much if we do streaming way or non streaming way.

nsivabalan commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java:
```diff
@@ -1096,6 +1119,126 @@ public void buildMetadataPartitions(HoodieEngineContext engineContext, List<Hood
+  public void completeStreamingCommit(String instantTime, HoodieEngineContext context, List<HoodieWriteStat> metadataWriteStatsSoFar, HoodieCommitMetadata metadata) {
+    List<HoodieWriteStat> allWriteStats = new ArrayList<>(metadataWriteStatsSoFar);
+    // update metadata for left over partitions which does not have streaming writes support.
+    allWriteStats.addAll(prepareAndWriteToNonStreamingPartitions(context, metadata, instantTime).map(writeStatus -> writeStatus.getStat()).collectAsList());
```
Note to reviewer: 
except RLI all other partitions that have been enabled will be written to MDT here.

danny0405 reviewed (changes requested):
block on my final review

nsivabalan commented:
@danny0405 @the-other-tim-brown @lokeshj1703 : rebased w/ latest master after we landed https://github.com/apache/hudi/pull/13286/

nsivabalan commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/BaseHoodieTableServiceClient.java:
```diff
@@ -316,33 +317,38 @@ protected HoodieWriteMetadata<O> compact(HoodieTable<?, I, ?, T> table, String c
     return compactionWriteMetadata;
   }
 
+  protected HoodieWriteMetadata<T> processWriteMetadata(HoodieTable table, HoodieWriteMetadata<T> writeMetadata, String instantTime) {
```
nope. we kept it to be generic since in other engines it does not do anything.

nsivabalan commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/BaseHoodieTableServiceClient.java:
```diff
@@ -316,33 +317,38 @@ protected HoodieWriteMetadata<O> compact(HoodieTable<?, I, ?, T> table, String c
+    Pair<List<HoodieWriteStat>, List<HoodieWriteStat>> dataTableAndMetadataTableHoodieWriteStats = triggerWritesAndFetchWriteStats(compactionWriteMetadata);
     // Fetch commit metadata from HoodieWriteMetadata and update HoodieWriteStat
-    CommonClientUtils.stitchCompactionHoodieWriteStats(compactionWriteMetadata, writeStats);
+    CommonClientUtils.stitchCompactionHoodieWriteStats(compactionWriteMetadata, dataTableAndMetadataTableHoodieWriteStats.getKey());
```
each of them are used just once. So, I left it as is.

nsivabalan commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/SparkRDDWriteClient.java:
```diff
@@ -123,23 +133,35 @@ public boolean commit(String instantTime, JavaRDD<WriteStatus> writeStatuses, Op
     if (canProceed) {
       // when streaming writes are enabled, writeStatuses is a mix of data table write status and mdt write status
       List<HoodieWriteStat> dataTableHoodieWriteStats = slimWriteStatsList.stream().filter(entry -> !entry.isMetadataTable()).map(SlimWriteStats::getWriteStat).collect(Collectors.toList());
       List<HoodieWriteStat> partialMetadataHoodieWriteStatsSoFar = slimWriteStatsList.stream().filter(entry -> entry.isMetadataTable).map(slimWriteStats -> slimWriteStats.getWriteStat()).collect(Collectors.toList());
```
I don't see major difference. we can leave it as is.

the-other-tim-brown commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/BaseHoodieTableServiceClient.java:
```diff
@@ -317,33 +318,40 @@ protected HoodieWriteMetadata<O> compact(HoodieTable<?, I, ?, T> table, String c
   public void commitCompaction(String compactionInstantTime, HoodieWriteMetadata<O> compactionWriteMetadata, Option<HoodieTable> tableOpt) {
     // dereferencing the write dag for compaction for the first time.
-    List<HoodieWriteStat> writeStats = triggerWritesAndFetchWriteStats(compactionWriteMetadata);
+    Pair<List<HoodieWriteStat>, List<HoodieWriteStat>> dataTableAndMetadataTableHoodieWriteStats = triggerWritesAndFetchWriteStats(compactionWriteMetadata);
```
I think it would be better for other devs to have a POJO here so it is clear which half of the pair is for the data table vs metadata table. If you disagree, can we at least introduce some local variables in the methods using the result that have data/metadata in the name so it is more clear what is passed to the other methods.

nsivabalan replied:
sure. Adding TableWriteStatsHolder

the-other-tim-brown commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/BaseHoodieTableServiceClient.java:
```diff
@@ -317,33 +318,40 @@ protected HoodieWriteMetadata<O> compact(HoodieTable<?, I, ?, T> table, String c
    * Commit Compaction and track metrics.
    */
-  protected void completeCompaction(HoodieCommitMetadata metadata, HoodieTable table, String compactionCommitTime) {
+  protected void completeCompaction(HoodieCommitMetadata metadata, HoodieTable table, String compactionCommitTime, List<HoodieWriteStat> metadataWriteStatsSoFar) {
```
nit on naming: `metadataWriteStatsSoFar` could be `currentMetadataWriteStats`?

nsivabalan replied:
hmmm. not sure if current suits here well. 
`metadataWriteStatsSoFar` signifies that this represents write stats to metadata written so far. 
or we can name `partialMetadataWriteStats`

the-other-tim-brown commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/BaseHoodieWriteClient.java:
```diff
@@ -1185,6 +1203,13 @@ public HoodieWriteMetadata<O> logCompact(String logCompactionInstantTime) {
+  /**
+   * Commit Log Compaction and track metrics.
+   **/
+  protected void completeLogCompaction(HoodieCommitMetadata metadata, HoodieTable table, String logCompactionCommitTime) {
```
This method is unused, is it still required?

the-other-tim-brown commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/io/HoodieMergeHandle.java:
```diff
@@ -270,7 +269,7 @@ protected void init(String fileId, Iterator<HoodieRecord<T>> newRecordsItr) {
       if (needsUpdateLocation()) {
         record.unseal();
-        record.setNewLocation(new HoodieRecordLocation(instantTime, fileId));
+        record.setNewLocation(newRecordLocation);
```
the `fileId` input to the method is unused, should it be removed now?

nsivabalan replied:
sg

the-other-tim-brown commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java:
```diff
@@ -1539,6 +1543,7 @@ protected void commitInternal(String instantTime, Map<String, HoodieData<HoodieR
     Pair<HoodieData<HoodieRecord>, List<HoodieFileGroupId>> result = tagRecordsWithLocation(partitionRecordsMap, isInitializing);
     HoodieData<HoodieRecord> preppedRecords = result.getKey();
     I preppedRecordInputs = convertHoodieDataToEngineSpecificData(preppedRecords);
+    List<HoodieFileGroupId> updatedMDTFileGroupIds = result.getValue();
```
This is unused, is that intentional?

nsivabalan replied:
removing it

the-other-tim-brown commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/HoodieMetadataWriteWrapper.java:
```diff
@@ -0,0 +1,127 @@
+public class HoodieMetadataWriteWrapper {
+
+  // Cached HoodieTableMetadataWriter for each action in data table. This will be cleaned up when action is completed or when write client is closed.
+  protected Map<String, Option<HoodieTableMetadataWriter>> metadataWriterMap = new ConcurrentHashMap<>();
```
can this be `private final`? Is it expected to undergo concurrent operations? if not, let's use a hashmap

nsivabalan replied:
will make it final. 
and also hashmap.

Initially, it was open and there could be concurrent accesses. 
But during iterations, I made getMetadataWriter() method in this class synchronous. 
So, we can make this hashmap.

the-other-tim-brown commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/HoodieMetadataWriteWrapper.java:
```diff
@@ -0,0 +1,127 @@
+/**
+ * Abstraction for data table write client and table service client to write to metadata table.
+ */
+public class HoodieMetadataWriteWrapper {
```
I think we can come up with a better name than `Wrapper`, maybe something that indicates this is for streaming writes? `StreamingMetadataWriteHandler` or something similar?

nsivabalan replied:
Open to suggestions. 
But I intentionally wanted to avoid "writeHandler" since its overlapping w/ "HoodieBackedTableMetadataWriter". 

`StreamingMetadataWritesManager` or `StreamingMetadataWritesDispatcher`

the-other-tim-brown commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/SparkRDDTableServiceClient.java:
```diff
@@ -36,17 +39,58 @@
 
 public class SparkRDDTableServiceClient<T> extends BaseHoodieTableServiceClient<HoodieData<HoodieRecord<T>>, HoodieData<WriteStatus>, JavaRDD<WriteStatus>> {
+
+  private HoodieMetadataWriteWrapper metadataWriterWrapper = new HoodieMetadataWriteWrapper();
```
make this `final`?

nsivabalan replied:
fixing it

the-other-tim-brown commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/HoodieMetadataWriteWrapper.java:
```diff
@@ -0,0 +1,127 @@
+   * Called by data table write client and data table table service client to perform streaming write to metadata table.
+   * @param table {@link HoodieTable} instance for data table of interest.
+   * @param dataTableWriteStatuses {@link WriteStatus} from data table writes.
+   * @param instantTime instant time of interest.
```
is this the instant time in the metadata table or data table?

the-other-tim-brown commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/SparkRDDWriteClient.java:
```diff
@@ -70,6 +70,7 @@ public class SparkRDDWriteClient<T> extends
     BaseHoodieWriteClient<T, JavaRDD<HoodieRecord<T>>, JavaRDD<HoodieKey>, JavaRDD<WriteStatus>> {
 
   private static final Logger LOG = LoggerFactory.getLogger(SparkRDDWriteClient.class);
+  private HoodieMetadataWriteWrapper metadataWriteWrapper = new HoodieMetadataWriteWrapper();
```
make this `final` as well?

nsivabalan replied:
sure

the-other-tim-brown commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/SparkRDDTableServiceClient.java:
```diff
@@ -36,17 +39,58 @@
+
+    if (isMetadataTable) {
+      ValidationUtils.checkArgument(dataTableWriteStats.isEmpty(), "For Metadata table,"
+          + "we do not expect any writes having WriteStatus referring to data table. ");
```
Nitpick: trim trailing space

nsivabalan replied:
taking care of it

the-other-tim-brown commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/SparkRDDTableServiceClient.java:
```diff
@@ -36,17 +39,58 @@
+    // When streaming writes are enabled, data table's WriteStatus is expected to contain all stats required to generate metadata table records and so each object will be larger.
+    // So, here we are dropping all additional stats and error records to retain only the required information and prevent collecting large objects on the driver.
+    List<SparkRDDWriteClient.SlimWriteStats> writeStatusMetadataTrackerList = tableServiceWriteMetadata.getWriteStatuses()
+        .map(writeStatus -> new SparkRDDWriteClient.SlimWriteStats(writeStatus.isMetadataTable(), writeStatus.getTotalRecords(), writeStatus.getTotalErrorRecords(),
```
This is similar to the code in SparkRDDWriteClient. We can expose a static method on `SlimWriteStats` like `static List<SlimWriteStats> from(JavaRDD<WriteStatus> writeStatuses)` to keep the code consistent between the two clients.

nsivabalan replied:
sure, good suggestion

the-other-tim-brown commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/SparkRDDTableServiceClient.java:
```diff
@@ -36,17 +39,58 @@
+
+  @Override
+  protected HoodieWriteMetadata<HoodieData<WriteStatus>> processWriteMetadata(HoodieTable table, HoodieWriteMetadata<HoodieData<WriteStatus>> writeMetadata, String instantTime) {
+    if (!isMetadataTable && config.isMetadataTableEnabled() && config.isMetadataStreamingWritesEnabled(table.getMetaClient().getTableConfig().getTableVersion())) {
```
This condition is the same as line 87 and I think there are similar checks in the SparkRDDWriteClient, is there a common place to put this logic to keep it consistent?

nsivabalan replied:
yes, there are minor difference in conditions we check in SparkRDDWriteClient and in the table service client. 
but I have tried to unify common denominator.

nsivabalan commented:
@the-other-tim-brown : addressed all of your feedback. have responded at places where I could not. 
I might need to check on some test failures. but have pushed out a commit. you can check it out.

nsivabalan commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/BaseHoodieTableServiceClient.java:
```diff
@@ -317,33 +318,48 @@ protected HoodieWriteMetadata<O> compact(HoodieTable<?, I, ?, T> table, String c
+    TableWriteStats tableWriteStats = triggerWritesAndFetchWriteStats(compactionWriteMetadata);
     // Fetch commit metadata from HoodieWriteMetadata and update HoodieWriteStat
-    CommonClientUtils.stitchCompactionHoodieWriteStats(compactionWriteMetadata, writeStats);
+    CommonClientUtils.stitchCompactionHoodieWriteStats(compactionWriteMetadata, tableWriteStats.getDataTableHoodieWriteStats());
```
nope. this is data table's commit metadata. which will get serialized to disk eventually. 
we should keep it clean and litter any metadata table related info in there.

nsivabalan commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/io/HoodieSparkFileGroupReaderBasedMergeHandle.java:
```diff
@@ -104,6 +104,8 @@ public void write() {
               row.getString(HoodieRecord.PARTITION_PATH_META_FIELD_ORD));
           HoodieSparkRecord record = new HoodieSparkRecord(recordKey, row, sparkSchema, false);
+          record.setCurrentLocation(newRecordLocation);
+          record.setNewLocation(newRecordLocation);
```
nope. new location is expected to be updated w/n MergeHandle. 
for eg: https://github.com/apache/hudi/blob/58191bdf5ffbf563386b69a1ac159e1436d47e57/hudi-client/hudi-client-common/src/main/java/org/apache/hudi/io/HoodieMergeHandle.java#L273

nsivabalan commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/HoodieMetadataWriters.java:
```diff
@@ -0,0 +1,126 @@
+/**
+ * Utilities for data table write client and table service client to write to metadata table.
+ */
+public class HoodieMetadataWriters {
```
I am not convinced this is a good name @danny0405  

How about HoodieMetadataWriteDelegate or HoodieMetadataWriteCoordinator ? 
@the-other-tim-brown : you also had some suggestions on the naming right.

danny0405 replied:
`XXXs` usually is a naming convention for utilities of `XXX`, which works for most of the cases.

the-other-tim-brown replied:
I have not heard of this convention before. My suggestion was [StreamingMetadataWriteHandler](https://github.com/apache/hudi/pull/13402#discussion_r2145857643) - I am not too picky either way but if this class is solely for the streaming write support then it should at least be mentioned in the javadoc if not the class name.

nsivabalan replied:
StreamingMetadataWriteHandler sounds good w/ me.

nsivabalan commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/TableWriteStats.java:
```diff
@@ -0,0 +1,54 @@
+/**
+ * Class to hold list of {@link HoodieWriteStat} for data table and metadata table.
+ */
+public class TableWriteStats {
```
1. we just have 2 args. not sure if builder might be overkill. 
2. sure. 
3. Table service invocation is slightly diff from Writeclient invocation. So, fromDataTableWriteMetadata() might only be useful for table service path. And also have to make this generic since this class has to deal with HoodieWriteMetadata<T> 
since we may not be able to unify across write client and table service client, not sure if this will help simplification. 
Within SparkRDDTableServiceClient, we ensure we just have one source of truth for this translation from HoodieWriteMetadata<JavaRDD<WriteStatus>> to List<HoodieWriteStat> dataTableHoodieWriteStats and private final List<HoodieWriteStat> metadataTableHoodieWriteStats (log compaction, compaction and clustering)

the-other-tim-brown replied:
Regarding (1) it is not about the number of args but rather the fact that the args have the same type and therefore can be easily swapped and only cause issues at runtime, making it harder to debug new code.

danny0405 reviewed (approved):
+1, I have no more comments.

nsivabalan commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java:
```diff
@@ -1462,7 +1470,7 @@ public void update(HoodieRollbackMetadata rollbackMetadata, String instantTime)
       HoodieInstant deltaCommitInstant = metadataMetaClient.createNewInstant(HoodieInstant.State.COMPLETED,
           HoodieTimeline.DELTA_COMMIT_ACTION, commitToRollbackInstantTime);
-      if (metadataMetaClient.getActiveTimeline().getDeltaCommitTimeline().containsInstant(deltaCommitInstant)) {
+      if (metadataMetaClient.getActiveTimeline().getDeltaCommitTimeline().containsInstant(deltaCommitInstant.requestedTime())) {
```
yes. looks like a mis during large refactoring.
